File: gaussian/gaussian_unknown_precision_model.R
evaluate_log_prior = function(inputs)
{
  return(dgamma(inputs$precision, 1, 1, log=TRUE))
}

simulate_prior = function()
{
  return(list(precision=rgamma(1, 1, 1)))
}

evaluate_log_likelihood = function(inputs, observed_data)
{
  return(sum(dnorm(observed_data$data, mean=0, sd=1/sqrt(inputs$precision), log=TRUE)))
}

simulate_model = function(inputs, observed_data)
{
  return(list(data = rnorm(length(observed_data$data), mean=0, sd=1/sqrt(inputs$precision))))
}

summary_statistics = function(observed_data)
{
  return(observed_data$data)
}

File: gaussian_unknown_precision_cpp/gaussian_unknown_precision_model.cpp
/***data***/
Data data()
{
  RandomNumberGenerator rng;
  rng.seed(1);
  size_t n = 100;
  arma::colvec sampled = rnorm(rng,n);
  Data data;
  data["y"] = sampled;
  return data;
}

/***evaluate_log_prior***/
double evaluate_log_prior(const Parameters &parameters)
{
  return dlnorm(parameters["tau"][0], 1.0, 1.0);
}

/***simulate_prior***/
Parameters simulate_prior(RandomNumberGenerator &rng)
{
  Parameters output;
  output["tau"] = rlnorm(rng, 1.0, 1.0);
  return output;
}

/***evaluate_log_likelihood***/
double evaluate_log_likelihood(const Parameters &parameters, const Data &data)
{
  double precision = parameters["tau"][0];
  return sum(dnorm(data["y"], 0.0, 1.0/sqrt(precision)));
}
